community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/MuninnPageCursor.java

```diff
@@ -97,10 +97,26 @@ public final void reset( MuninnPage page )
     @Override
     public final boolean next( long pageId ) throws IOException
     {
+        if ( isAlreadyPositionedAt( pageId ) )
+        {
+            return true;
+        }
         nextPageId = pageId;
         return next();
     }
 
+    /**
+     * Checks if this cursor is already positioned at the page with given id.
+     *
+     * @param pageId the page id to check
+     * @return <code>true</code> if the current page id is the same as the given one plus current page has not been
+     * evicted, <code>false</code> otherwise.
+     */
+    private boolean isAlreadyPositionedAt( long pageId )
+    {
+        return currentPageId == pageId && page.isBoundTo( pagedFile.swapper, currentPageId );
```

Comment:
You might be able to simplify this to `currentPageId == pageId & pagedFile != null` because `shouldRetry` will protect the read cursors, and write-locked pages cannot be evicted. The `pagedFile != null` part would guard for calling this method after the cursor has been closed. And using bit-wise AND will save us some branching logic in the generated code.